Feedback 1: Plexus folks on invariance and DMI blogs

mokyen opened the pull request:
I have finished the shorter of my two blogs about Decision Making Isolation. I would appreciate feedback from you if you’re willing and able. 

I am writing a blog about DMI, but one of the key concepts is that of ‘invariance’. This topic seemed too big to be included in the DMI blog, so I broke it out.

These blogs will serve as a direct input as I build a presentation on this topic in the future. I expect to pull from the terminology, code examples, analogies, etc. The feedback I’m looking for includes:

•	Do you agree with the technical concepts?
•	Are the concepts clear?
•	Does the terminology make sense?
•	Are the coding examples clear and useful?
•	What do you think C++ experts will question or criticize?

https://github.com/mokyen/okyensourcesoftware/blob/main/invariance.md

Note that this blog does link to the main DMI blog, but I’d prefer if you didn’t dig into that too much until it’s a little further along. Additionally, I believe the grammar is in decent shape, but that isn’t a big point of emphasis right now. (I might go ahead and publish to my website soon, so I’ll take feedback on the syntax and grammar if it jumps out at you.)

mokyen committed:
Update Started DMI blog

Just beginning

mokyen committed:
Added some definitions of SRP

mokyen committed:
Update dmi.md. Added in DMI details

mokyen committed:
Added invariance.md

mokyen committed:
Update dmi.md. Added in cleaner version of the CppCoreGuidelines. Also updated DMI types a bit and references

mokyen committed:
Update dmi.md. Added author of hexagonal architecture

mokyen committed:
Update dmi.md with entities terminology

mokyen committed:
Update dmi.md. Started adding the key takeaways from CppCG. Changed formatting of those takeaways

mokyen committed:
Update dmi.md. Added reference to Klaus's quote about quality software

mokyen committed:
Update dmi.md. Filled in entities definitions.

mokyen committed:
Update dmi.md Added key takeAways from core guidelines

mokyen committed:
Update dmi.md. Added link to OCP Robert Martin article for reference

mokyen committed:
Update invariance.md

mokyen committed:
Updated gitignore for mkdocs.

mokyen committed:
Adding mkdocs stuff.

mokyen committed:
Added in CNAME file.

mokyen committed:
Update dmi.md. Updated intro to be less absolute.

mokyen committed:
Started actually writing out DMI. On a good roll with examples from SESG presentation!

mokyen committed:
Update dmi.md. continued discussion of the first refactored example

mokyen committed:
Update dmi.md. Added definition of quality software

mokyen committed:
Added CNAME to docs directory in hopes of not overriding the external domain on every push.

mokyen committed:
Expanded the DMI section.
Added more about the RequestHandler simple example.
Added whole section about unit testing in DMI
Expanded section about clean architecture

mokyen committed:
Added about-me page.

mokyen committed:
Cleanup of invariance.md. Changed title of homepage.

mokyen committed:
Significant progress on invariance post. Broke down invariance into three types. Still need to fill in a bit better on at least the last one.

mokyen committed:
Added navigation specificaiton to the mkdocs.yml. Added about page.

mokyen committed:
Fixed some markdown issues in invariance blog.

mokyen committed:
Update invariance.md. Minor typo resolutions.

mokyen committed:
Update dmi.md. Added a vision for DMI and updated the goals

mokyen committed:
Update invariance.md. Updated section about invariance type for property of a member

mokyen committed:
Fixed gramatical and syntax issues in invariance.md

mokyen committed:
Fixed the code examples so they are working code. One issue exists with the ohms law example.

mokyen committed:
More code example formatting changes.

mokyen committed:
Ran code examples through godbolt clang-format. More code example formatting changes.

mokyen committed:
Finished writing invariance blog. Last two sections need to finish going through grammarly.

mokyen committed:
Finished last few paragraphs of the invariance blog. It's ready for review!

mokyen committed:
Update invariance.md. Added notes about how DMI is where application is discussed

natillet commented on invariance.md:
```diff
@@ -0,0 +1,204 @@
+> An invariant is a logical condition for the members of an object that can be expected to be true from the time the constructor exits to the time the destructor is called. These logical conditions fall into three categories.
+>
+> * A relationship between at least two data members such that all members cannot vary independently.
+> * A relationship between a data member and a resource (aka RAII).
```
I find the mention of RAII here a bit confusing. It almost feels like it's saying the category is RAII, and that doesn't seem right.

mokyen replied:
RAII is a category, but I misunderstood that the invariance is the guarantee that RAII enforces. I clarified this

natillet commented on invariance.md:
````diff
@@ -0,0 +1,204 @@
+};
+```
+
+The variable `m_isLightOn` on can change completely independently from `m_brightness`, but the reverse is not true. Therefore, these two pieces of data still have an invariance, or, said another way, the relationship between these two is an invariant.
````
```suggestion
The variable `m_isLightOn` on can change completely independently from `m_brightness`, but the reverse is not true. Therefore, these two pieces of data still have an invariance, or, said another way, the relationship between these two contains an invariant.
```

To my understanding of `invariant`, it is the thing that cannot change, so it feels wrong to say the relationship in this case _is_ the invariant. I think it's more accurate to say the relationship _contains_ an invariant. Or maybe there's another way to put it that's better, but I haven't thought of it yet.

mokyen replied:
I removed this last clause and just stated the invariance itself.

natillet commented on invariance.md:
```diff
@@ -0,0 +1,204 @@
+
+The variable `m_isLightOn` on can change completely independently from `m_brightness`, but the reverse is not true. Therefore, these two pieces of data still have an invariance, or, said another way, the relationship between these two is an invariant.
+
+## Relationship Between a Member and Resource
```
I think I just don't get this section. Are we saying that owning a resource equates to establishing an invariant? Is the invariant that the owned resource exists? This section seems like a decent description of RAII, but doesn't tie it to invariance for me.

mokyen replied:
Overhauled this whole section based on my new view on the invariance

natillet commented on invariance.md:
```diff
@@ -0,0 +1,204 @@
+
+> Resource Acquisition Is Initialization, or RAII, is a C++ programming technique that binds the life cycle of a resource that must be acquired before use (allocated heap memory, thread of execution, open socket, open file, locked mutex, disk space, database connection—anything that exists in limited supply) to the lifetime of an object.
+
+RAII establishes a relationship between the data member and the resource. Placing a file handle as a public member would be dangerous, as changing it could create a memory leak. RAII creates an invariance between the member and the resource.
```
This is another instance of mentioning public member that I assume is a reflection on the class vs struct. I had a hard time with why this was being pointed out initially.

mokyen replied:
I called out how this necessitates a class

natillet commented on invariance.md:
```diff
@@ -0,0 +1,204 @@
+
+Here, the invariance is a condition between a member and the constant *2*.
+
+I admit that this third category could be considered broad and open-ended. Developers could choose to interpret this to justify just about any data needing to be in a class. Most of the time, the manipulation of a data member is going to be compared to some constant data or literal. Developers should scrutinize if a piece of data needs that rule enforced before choosing to use a class.
```
Reading this I had a hard time deciding if "Developers" is referring to writers or users of the class with the invariant. The last sentence makes me think that it's users of the class. But you also started the paragraph in a way that I figured, as a consumer of this article, likely considering _making_ classes, that I should be reading this as a writer of classes.

mokyen replied:
I felt uncomfortable with this section and that this was just too broad and open-ended. This third type of invariance is the newest one for me, and I'm struggling to come up with more concrete guidance. I left a TODO

natillet commented on invariance.md:
```diff
@@ -0,0 +1,204 @@
+
+Here, the invariance is a condition between a member and the constant *2*.
+
+I admit that this third category could be considered broad and open-ended. Developers could choose to interpret this to justify just about any data needing to be in a class. Most of the time, the manipulation of a data member is going to be compared to some constant data or literal. Developers should scrutinize if a piece of data needs that rule enforced before choosing to use a class.
```
I don't understand `Developers could choose to interpret this to justify just about any data needing to be in a class.`

mokyen replied:
Removed this entirely

natillet commented on invariance.md:
```diff
@@ -0,0 +1,204 @@
+
+## 'Invariant' Doesn't Mean Immutability
+
+Keep in mind that invariance means a relationship, not that the data cannot change. The term in software for something that cannot change is immutability.
```
```suggestion
Keep in mind that invariance describes a relationship, not that the data cannot change. The term in software for something that cannot change is immutability.
```

mokyen replied:
I rephrased this and made it more explicit

natillet commented on invariance.md:
```diff
@@ -0,0 +1,204 @@
+
+With these simple examples in mind, here is what the Guidelines say about when to use a class in [C.2](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c2-use-class-if-the-class-has-an-invariant-use-struct-if-the-data-members-can-vary-independently):
+
+> ### <a name="Rc-struct"></a>C.2: Use `class` if the class has an invariant; use > `struct` if the data members can vary independently
```
```suggestion
> ### <a name="Rc-struct"></a>C.2: Use `class` if the class has an invariant; use `struct` if the data members can vary independently
```

natillet commented on invariance.md:
```diff
@@ -0,0 +1,204 @@
+
+> volts = current * resistance
+
+Ohm's law dictates that if you vary any one of the parameters, at least one other must also change. If the voltage increases, either the current or resistance must also increase. If the current decreases, either the resistance will increase or the voltage will decrease. They cannot be changed independently. If we were to code up an object that held these parameters, the parameters couldn't be pub ...
```
You bring up that the parameters shouldn't be public members here, which is true, but also it took me a bit to get that your unspoken point was that they shouldn't be in a struct. Because the topic is "invariance", and it was a take off from the class vs struct, I wasn't in the headspace to be reflecting on this as an angle.

mokyen replied:
I added a section at the beginning that aims to clarify the goal in understanding invariance.

natillet commented on invariance.md:
```diff
@@ -0,0 +1,204 @@
+
+## But what about when my invariants or data change?
+
+One of the concerns that can arise from choosing to put data into a struct instead of a class with getters and setters is that those methods provide an interface that insulates callers if the underlying data needs to change. I agree that this not only can but absolutely *will* happen at some point. However, I think a couple of things can serve as protections. First, [DMI](dmi.md) gives guidance t ...
```
This first sentence is confusing to me. It sounds like "one of the concerns [...] is that those methods provide an interface that insulates callers if the underlying data needs to change" but the choice was putting "data into a struct instead of a class". So it reads as the problem is the struct because the class has methods that insulate callers from change, and I don't get that. From there I don't know what point the rest of the paragraph is making until we get to your "second" point, which suggests you're advocating converting structs to classes.
I don't know that this answers the question in the heading, and I don't have in my head a thought of a situation where that might occur off hand.

mokyen replied:
I can completely see where I wasn't clear in my wording. I rewrote this as a more explicit problem paragraph and left most of the solution the same.

mokyen committed:
typo

Co-authored-by: Elene <natille@gmail.com>

mokyen reviewed (commented):
Responded to Elene's commentary

mokyen committed:
Greatly modified the invariance blog based on Elene's feedback.

mokyen committed:
Update invariance.md. Updated third type of invariance to discuss classes with algorithms instead of just const comparisons.

natillet commented on invariance.md:
````diff
@@ -0,0 +1,222 @@
+};
+```
+
+Here, the invariance is a condition between a member and the constant *2*.The most common case of this type of invariance is comparisons to a constant or literal, like the previous example. However, a similar case occurs when the invariance is a mathematical algorithm that must be applied to a member. For instance, a class could exists that stores an angle in radians but must convert the value to ...
````
```suggestion
Here, the invariance is a condition between a member and the constant *2*. The most common case of this type of invariance is comparisons to a constant or literal, like in this example. However, a similar case occurs when the invariance is a mathematical algorithm that must be applied to a member. For instance, a class could exist that stores an angle in radians but must convert the value to be in the range -2π to 2π. This is more involved than simply comparing to a constant, but it's a similar logical assertion that necessitates using a class.
```

natillet commented on invariance.md:
```diff
@@ -0,0 +1,222 @@
+
+With these simple examples in mind, here is what the Guidelines say about when to use a class in [C.2](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c2-use-class-if-the-class-has-an-invariant-use-struct-if-the-data-members-can-vary-independently):
+
+> ### <a name="Rc-struct"></a>C.2: Use `class` if the class has an invariant; use `struct` if the data members can vary independently
```
```suggestion
> ### C.2: Use `class` if the class has an invariant; use `struct` if the data members can vary independently
```

Maybe doesn't matter, but this could be removed.

natillet commented on invariance.md:
```diff
@@ -0,0 +1,222 @@
+> * A guarantee that any function with access to a resource may be called at any time, i.e. implementing RAII.
+> * A property of a data member that must always be true.
+
+A key for me is remembering that an invariance is **a logical condition**.
```
I like some of the discussion on this thread: https://softwareengineering.stackexchange.com/questions/32727/what-are-invariants-how-can-they-be-used-and-have-you-ever-used-it-in-your-pro

A couple of the things that caught my attention in the discussion:

* [citation](https://softwareengineering.stackexchange.com/a/364286)

  > An invariant is any logical rule that must be obeyed throughout the execution of your program that can be communicated to a human, but not to your compiler.

  This speaks to me because it ties in that you are designing in some constraints that may be important enough to validate at specified points during execution.

  Thinking about this, you could expand this sentence to say something like "... an invariance is **a logical condition** required of the class' purpose".

* I also like the example in the same comment regarding the gear shift. Perhaps there's another example that's actually software specific, but for the gear shift example, thinking about a situation where you design a class to represent it that treats each gear independently would require lots of invariant checks as opposed to the design where the single gear shift knob selects one and can only select one gear which has the same invariance but with less verification required. Maybe this is really out of scope the more I think about it, but it's an interesting perspective to think through.

natillet commented on invariance.md:
```diff
@@ -0,0 +1,222 @@
+
+The variable `m_isLightOn` on can change completely independently from `m_brightness`, but the reverse is not true. Therefore, these two pieces of data still have an invariance: the brightness will only be updated when the light is on.
+
+## Guaranteeing Manipulation of a Resource, or RAII
```
The term "manipulation" doesn't feel right here. "Guaranteeing lifetime of an owned resource", maybe?

mokyen replied:
Changed per your suggestion

natillet commented on invariance.md:
```diff
@@ -0,0 +1,222 @@
+
+## Guaranteeing Manipulation of a Resource, or RAII
+
+Resources such as files, sockets, or heap memory need to be managed so that the right steps are made to open/close, connect/close, or allocate/deallocate these assets. It can be dangerous to assume that a caller will properly implement the correct initialization/deinitialization, so the [Resource Allocation is Initialization, or RAII,](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines# ...
```
```suggestion
Resources such as files, sockets, or heap memory need to be managed so that the right steps are made to open/close, connect/close, or allocate/deallocate these assets. It can be dangerous to assume that a caller will properly implement the correct initialization/deinitialization, so the [Resource Allocation is Initialization, or RAII,](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#p8-dont-leak-any-resources) technique was devised to encapsulate the assets. When implementing RAII, the wrapping entity (object) guarantees that any function that can access the underlying resource can be called from the end of the object's constructor until the object's destructor is called.
```

natillet replied:
Another thought might be that the invariant here is the accessibility/usability of the resource.

natillet commented on invariance.md:
```diff
@@ -0,0 +1,222 @@
+
+Since the goal of RAII is to insulate the resource from direct manipulation, this necessitates that we use a class instead of a struct based on our previous discussion.
+
+*On a side note, I really prefer the term CADRe, or Constructor Acquires-Destructor Releases, as a better name for remembering RAII. It is just easier to remember the meaning. However, it's hard to change a term that's been around longer than the Super Nintendo.*
```
I'm sorry, I had to look this up. [Super Nintendo (SNES)](https://en.wikipedia.org/wiki/Super_Nintendo_Entertainment_System) was released in Japan in 1990 and US in 1991. It seems that RAII as a name for this was introduced in 1994, even though the technique existed since at least the 80s.

mokyen replied:
Hmm. I might have to rephrase this. My reference was the link below, but I can have some fun with this.

https://www.stroustrup.com/bs_faq.html#:~:text=I%20held%20back%20the%20introduction%20of%20exceptions%20into%20C%2B%2B%20for%20half%20a%20year%20until%20I%20discovered%20RAII%20(1988).%20RAII%20is%20an%20integral%20and%20necessary%20part%20of%20the%20C%2B%2B%20exception%20mechanism.

natillet commented on invariance.md:
```diff
@@ -0,0 +1,222 @@
+
+* [C.8](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c8-use-class-rather-than-struct-if-any-member-is-non-public): Use `class` rather than `struct` if any member is non-public
+
+While `class` and `struct` are relatively equivalent in C++, the Guidelines set the convention that structs are intended to only contain public members. Because all public data members can be directly modified, I believe it can be inferred that structs also shouldn't have public member functions. In the [DMI blog](dmi.md), we'll go more into the preference for free functions. For now, we'll assum ...
```
Might be nice to define the initialism DMI here at its first use.

mokyen replied:
Defined per your suggestion

natillet commented on invariance.md:
```diff
@@ -0,0 +1,222 @@
+
+## But what about when my invariants or data change?
+
+One potential critique of these guidelines for choosing a struct versus a class is that public member functions provide an interface that isolates the private data members from the users of that object. If I have a class with a getter, I can change the type of that private data without modifying calls to the getter as long as the return type remains the same. If a struct is used and the data memb ...
```
```suggestion
One potential critique of these guidelines for choosing a struct versus a class is that public member functions provide an interface that isolates the private data members from the users of that object. If I have a class with a getter, I can change the type of that private data without modifying calls to the getter as long as the return type remains the same. If a struct is used and the data members are manipulated directly, any change to the types of those data members impacts the struct's users. Additionally, if a struct is used and we need to introduce an invariance (which would necessitate private data and therefore a switch from a struct to a class), then we'll also have to modify all the users. Using a class is less likely to cause the callers of its member functions to change.
```

natillet commented on invariance.md:
```diff
@@ -0,0 +1,222 @@
+
+One potential critique of these guidelines for choosing a struct versus a class is that public member functions provide an interface that isolates the private data members from the users of that object. If I have a class with a getter, I can change the type of that private data without modifying calls to the getter as long as the return type remains the same. If a struct is used and the data memb ...
+
+I agree that this is definitely a tradeoff. I think a couple of things can serve to minimize the impact of these scenarios. First, [DMI](dmi.md) gives guidance that keeps the scope of data small, meaning that the scope of updates when objects change will also be small. Second, changing from setting a variable directly to calling a getter or setter when a new invariance is introduced is a minor re ...
```
Not being familiar with DMI, I don't know whether it covers classes or structs, so I don't know in which context to frame the guidance. The "First, DMI..." sentence leads me to assume it's regarding structs. The "Second, ..." sentence throws me off because now we're talking about refactoring and I don't even know for sure from what to what yet. The "Third, ..." sentence contributes very little to my understanding; I can think about it and infer that refactoring is part of coding life. "The setter/getter" sentence feels like an after though that doesn't connect to the last sentence's "these types of refactors" maybe only because referring to the "setter/getter interface" as "fluid" doesn't clarify to me what you mean entirely.

mokyen committed:
Cleaned up the DMI blog for typos and grammar.

mokyen commented on invariance.md:
```diff
@@ -0,0 +1,222 @@
+
+Since the goal of RAII is to insulate the resource from direct manipulation, this necessitates that we use a class instead of a struct based on our previous discussion.
+
+*On a side note, I really prefer the term CADRe, or Constructor Acquires-Destructor Releases, as a better name for remembering RAII. It is just easier to remember the meaning. However, it's hard to change a term that's been around longer than the Super Nintendo.*
```
```suggestion
*On a side note, I really prefer the term CADRe, or Constructor Acquires-Destructor Releases, as a better name for remembering RAII. It is just easier to remember the meaning. However, it's hard to change a term that might have been around [longer than the Super Nintendo](https://www.stroustrup.com/bs_faq.html#:~:text=I%20held%20back%20the%20introduction%20of%20exceptions%20into%20C%2B%2B%20for%20half%20a%20year%20until%20I%20discovered%20RAII%20(1988).%20RAII%20is%20an%20integral%20and%20necessary%20part%20of%20the%20C%2B%2B%20exception%20mechanism) (or at least [older than the Nintendo 64](https://www.cardinalpeak.com/blog/raii-introduction)) .*
```

mokyen committed:
Update invariance.md. Applied suggestion

Co-authored-by: Elene <natille@gmail.com>

mokyen committed:
Apply suggestions from code review

Co-authored-by: Elene <natille@gmail.com>

mokyen committed:
Updated DMI blog. Filled in the SRP and OCP intro sections. Filled in more details on the THree Types of Code sections. Still not happy with the IO section. Made a few changes per review comments to invariance.

mokyen committed:
Added more to the dmi three types of code sections. Added in TODO sections for all remaining topics. Started in on a sudoku example.

mokyen committed:
Update dmi.md. Added rough example for wiring about how it meets the Clean Code definition

mokyen committed:
Update dmi.md. Filled out the Architectural boundaries section as a stream of consciousness via voice to text. really rough

mokyen committed:
Update dmi.md. started section on invariance. rough stream of consciousness via voice to text

mokyen committed:
Update dmi.md. Cleaned up the Invariance and Arch Boundary sections. Still might need to be expanded, but in decent shape

mokyen committed:
Update dmi.md. Incorporated mux of the feedback from Matt. Added section on free functions

mokyen committed:
Update dmi.md. Added Note about the virtual functions actually being referenced in the enforcement part of a guideline and fixed a typo

mokyen committed:
Updated Quality Software to soften language around adding readability to Klaus defintion. Added section detailing the motivation for creating DMI. Cleaned up sentence about OCP in the initial DMI example. Started section about my design process. Still need a bigger code example.

mokyen committed:
Added citation for Scott Meyers Effective C++ in the free function references section.

mokyen committed:
Update dmi.md. Fixed a bad link

mokyen committed:
Update dmi.md. Added to unit test cost

mokyen committed:
Update dmi.md. Incorporated the feedback from Matt on the three types of code

mokyen committed:
Update dmi.md. Changed rule about unit testing in DMI to say it should be tested thoroughly. Added quote to Wiring section from clean code about well-written prose

mokyen committed:
Update invariance.md. Updated first 2/3 with content from presentation. Need to consider if the date example is used here or in the DMI blog.

mokyen committed:
Update invariance.md. Added the Reason from the guidelines about struct. Changed a few key words to get ready for publishing the blog. Made references to DMI talk a out the future. Still undecided about adding the date example

mokyen committed:
Updated .gitignore for the accidental double venv that I installed. Added a page on the website for the website itself. Added a requirements.txt for the dependencies.

mokyen committed:
Publishing the invariance blog!

mokyen committed:
Fixed name of invariance blog file. Renamed the DMI blog page to draft so I can create a placeholder on the actual site.

mokyen committed:
Added a WIP dmi.md blog post. Fixed an issue in invariance blog for the Important alert that isn't supported outside of github right now. Website is live!

mokyen committed:
Update invariance.md. Renamed the alert from IMPORTANT to KEY POINT

mokyen committed:
Updated the Key Point to be a sparkle. Updated DMI blog draft to match, but it might not be necessary.

mokyen committed:
Updated homepage to remove WIP. Added links to presentations to BIO.

mokyen committed:
Adding new blog

mokyen committed:
Update godbolt_cmake_gtest.md

Added updates/revision from claud. Needs more work

mokyen committed:
Update mkdocs.yml

Added godbolt cmake blog to the navigation

mokyen committed:
Update index.md

Updated homepage with godbolt cmake blog

mokyen committed:
Update godbolt_cmake_gtest.md

Finished updates. Ready for merge!

mokyen committed:
Merge pull request #4 from mokyen/godbolt_cmake_gtest

Godbolt cmake gtest

mokyen committed:
Update godbolt_cmake_gtest.md

Updated link for the example project

mokyen committed:
Merge pull request #5 from mokyen/fixing_example_link

Update the link to the example

mokyen committed:
Create dmi_blog_feedback.txt

mokyen committed:
Update bio.md. Added link to CppNow 2024 lightning talk

mokyen committed:
Update bio.md with DMI talk link

mokyen committed:
Blog comments and rss feed (#6)

* Update requirements.txt

Added to requirements 
mkdocs-blogging-plugin
mkdocs-rss-plugin

* Update mkdocs.yml

Updated the mkdocs.yml to add RSS feed and blog

* Update mkdocs.yml

Added in JavaScript for the commenting plugin

* Create Temp.txt

* Rename docs/godbolt_cmake_gtest.md to docs/blog/godbolt_cmake_gtest.md

Moved to blog folder

* Rename docs/invariance.md to docs/blog/invariance.md

Moved to blog folder

* Update and rename docs/dmi.md to docs/blog/dmi.md

Moved to blog folder

* Create deploy.yml

Added deploy script on PR creation

* Update mkdocs.yml

Removed JavaScript to hopefully fix deploy preview issue

* Update deploy.yml

Hopefully fixed deploy wrror

* Update deploy.yml

changed the links to dynamically redirect to the preview site for preview builds

* Update deploy.yml

added printing a link to the preview site

* Update deploy.yml

updated deploy configuration to check if a preview branch already exists before creating a new branch

* Update deploy.yml

Fixed syntax error

* Update deploy.yml

another attempt at fixing the syntax issue

* Update deploy.yml

Another syntax fix attempt

* Update deploy.yml

* Update deploy.yml

different attempt at fixing deploy

* Update deploy.yml

Another attempt at fixing syntax

* Update deploy.yml

Trying to fix issue with finding if the branch already exists.

* Update mkdocs.yml

Added the site_url

* Update deploy.yml

Updated to hopefully better fix the preview site actually existing.

* Update deploy.yml

Changed method for creating a preview to stop fighting the settings for the preview

* Update deploy.yml

Instead of trying to serve previews from different paths on your main GitHub Pages site, we could publish each preview to its own independent GitHub Pages URL pattern.

* Rename docs/blog/Temp.txt to docs/blog/posts/Temp.txt

Creating posts directory

* Update and rename docs/blog/dmi.md to docs/blog/posts/dmi.md

* Update and rename docs/blog/godbolt_cmake_gtest.md to docs/blog/posts/godbolt_cmake_gtest.md

Moved to align with new structure for blogging plugin

* Update and rename docs/blog/invariance.md to docs/blog/posts/invariance.md

Updated to align with blogging plugin structure

* Create _index.md as a landing page for the blog

* Update mkdocs.yml

Updated the blog structure

* Update and rename index.md to _index.md

Renamed and hopefully fixed relative links

* Update and rename _index.md to index.md

Updated paths again

* Update deploy.yml

Updated fetch depth to help fix build issue

* Update mkdocs.yml

Updated to fix build issue with author not found

* Update mkdocs.yml

Another attempt at author build failure

* Update mkdocs.yml

Reverting changes and making author mokyen

* Update _index.md

Updated author

* Delete docs/blog/posts/Temp.txt

* Update dmi.md

* Update godbolt_cmake_gtest.md

* Update invariance.md

* Update mkdocs.yml

Updated RSS config

* Update _index.md

Updated blog landing page.

* Update mkdocs.yml

Updated navigation for blog to point to landing page

* Update _index.md

Updated syntax for different post features

* Update deploy.yml

Added back in the giscus stuff

* Update deploy.yml

Added code for comments plugin

* Create comments.html

Added an html page for comments.

* Update dmi.md

Enabled comments

* Update deploy.yml

Reverted accidental changes to deploy that were supposed to be in mkdocs.yml

* Update mkdocs.yml

Updated with the overrides for comments.

* Update mkdocs.yml

* Update dmi.md

* Update godbolt_cmake_gtest.md

* Update invariance.md

* Update godbolt_cmake_gtest.md

* Update invariance.md

* Update _index.md

Hopefully fixing the blog index page to display blogs

* Update _index.md

Simplified blog page

* Update _index.md

Removed header blob in attempt to get blog to populate

* Update _index.md

* Update _index.md

mokyen committed:
Fix rss (#7)

* Update deploy.yml

Updated to use requirements.txt instead

* Update ci.yml

Updating to get other dependencies in requirements.txt

* Update requirements.txt

added in missing dependencies

mokyen committed:
Fixing blog index page (#8)

* Update and rename _index.md to index.md

Renaming and Simplifying for testing

* Update mkdocs.yml

Renamed index.md in the Nav section to match file rename

* Update index.md

Reverted to old data

* Update index.md

Removed the info on latest blogs in favor of auto stuff

mokyen committed:
Update godbolt_cmake_gtest.md

Added tag to separate the blog's excerpt

mokyen committed:
Update invariance.md

Added tag to define the blog excerpt

mokyen committed:
Fix comments not loading (#9)

* Create main.html

Added a theme page to try to get comments working

* Update mkdocs.yml

Trying something different for comments

* Merging in main (#10)

* Update godbolt_cmake_gtest.md

Added tag to separate the blog's excerpt

* Update invariance.md

Added tag to define the blog excerpt

* Update index.md

Renamed to just blog

* Update index.md

mokyen committed:
Update main.html (#11)

* Update main.html

Added RSS feed link on the bottom of every blog page

* Update main.html

Simplified RSS feed link

* Update main.html

Reverting add of RSS feed

* Update main.html

Another way for RSS

* Create footer.html

Adding RSS feed

* Update main.html

moved RSS code outside of the conditional block for comments

* Update footer.html

* Update footer.html

Made RSS feed links relative to fix them in the preview site

* Update main.html

mokyen committed:
Cppnow24 lightningtalk (#14)

* Create debugging_like_house.md

* Update debugging_like_house.md

mokyen committed:
Created cppnow_what_to_know.md (#15)

* Create cppnow_what_to_know.md

* Update cppnow_what_to_know.md

* Update cppnow_what_to_know.md

* Update cppnow_what_to_know.md

* Update cppnow_what_to_know.md

Added note about the slides being posted.

* Update cppnow_what_to_know.md

* Update cppnow_what_to_know.md

* Update cppnow_what_to_know.md

* Update index.md

* Update cppnow_what_to_know.md

mokyen committed:
Update ci.yml

mokyen committed:
Update cppnow_what_to_know.md

mokyen committed:
Update mkdocs.yml
